### ISSUE TITLE
Generate Site ID

### ISSUE THREAD

**Issue by nathankota** [uic-utah/uic-inventory#259]
- [x] Generate a Site ID during the site creation workflow once the site information is complete (contacts, info, location)
- [x] The Site ID must not change if any part of a site is edited at any time by an admin or the applicant
- [x] Display the Site ID in the Site Information panel:
![image](https://github.com/uic-utah/uic-inventory/assets/15659642/77da0458-15b8-4612-acf0-b3f2bb3fb91d)
- [x] Display the Site ID on the Inventory Review page
- [x] Display the Site ID on the Inventory Review report

### Site ID format:
- UT**U**ccSXXXXXXXX, where
- UT**U** is a constant
- cc is the county FIPS code (2 digits)
- S is a constant (i.e. "Site")
- XXXXXXXX is a randomly generated, unique sequence of eight (8) alpha-numeric characters

**Comment by steveoh** [comment]
The attribute rules may make data writing from outside of pro an issue. Someone will need to test this.

**Comment by rsparker-utah** [comment]
One quick edit to the Site ID format:
- UT**U**ccFXXXXXXXX, where
- UT**U** is a constant
- cc is the county FIPS code (2 digits)
- F is a constant
- XXXXXXXX is a randomly generated, unique sequence of eight (8) alpha-numeric characters

In regards to the attribute rule for the FacilityID, it looks to be triggered immediately during the creation of the Facility polygon. I'm not sure what this would mean in regards to writing to the FacilityID from outside Pro, but after creation (at least within ArcPro) the FacilityID field is not editable (similar to the OBJECTID or GUID fields).

**Comment by nathankota** [comment]
@rsparker-utah
- I have updated the original comment to include your Site ID (i.e. FacilityID) specifications
- Can you install another database tool such as [DBeaver](https://dbeaver.io/) to test editing outside of Pro sooner than later? It will be important to know all of the nuances of the attribute rules that may affect editing the database, like the one you described above. As the primary user, you're well poised to test and catalog those nuances so we can plan for/around them.

**Comment by rsparker-utah** [comment]
Adding [Link](https://github.com/uic-utah/uic-attribute-rules/blob/master/src/rules/arcade/facility/idCalculation.js) to FacilityID calculation attribute rule for reference.

**Comment by steveoh** [comment]
@rsparker-utah During your testing did you verify that you can write a constant FacilityID that will persist updates to the records geometry or whatever events that may trigger it to change?

**Comment by rsparker-utah** [comment]
@steveoh From what I can tell, yes. My understanding is that the [code](https://github.com/uic-utah/uic-attribute-rules/blob/master/src/rules/arcade/facility/idCalculation.js) that generates the FacilityID is only triggered during the creation of the Facility polygon within ArcPro. Once the FacilityID is created, it becomes an uneditable field. 

When I tested writing to the database using SQL Server Management Studio (SSMS), the value I provided for the FacilityID field seemed to be accepted fine by the database. I was then able to view this created record in ArcPro and it maintained the same uneditable properties in ArcPro as other records that were created using ArcPro:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/50fba247-41a1-42b1-83ee-40352424b66c)
It seems to me that writing directly to the database using an outside program like SSMS bypasses the trigger that executes the code that creates the FacilityID. Does that help at all regarding your question?

**Comment by steveoh** [comment]
What happens if you wrote a facility id to a record with SSMS with the value of `test`. Then use pro to edit the facility in some way, for instance move the polygon to a different county or something etc, does the facility id of `test` persist those edits?

**Comment by rsparker-utah** [comment]
I tried moving the polygon and no, the facility id of `test` does not persist when moving the facility to a different county within ArcPro.

Here are the steps I took to test this for reference:

1.) I wrote a new record to the facility table in SSMS using the following query:
```
INSERT INTO UDEQ.UICADMIN.UICFacility_evw (OBJECTID, GUID, CountyFIPS, NAICSPrimary, FacilityID, FacilityName, FacilityAddress, FacilityCity, FacilityZIP, Comments, CreatedOn, ModifiedOn, EditedBy, GlobalID, Shape, CreatedBy)
VALUES (35441, 'A4DE0737-1144-4C13-95CC-B3C4170C8B36', 49003, '0', 'test', 'SSMS FacilityID Test', '500 GSL St.', 'Promontory', '84329', 'Facility created to test behavior of FacilityID field when editing database outside of ArcPro.',  '8/3/2023  5:34:18 PM', '8/3/2023  5:34:18 PM', 'SSMS', 'F77AD020-D067-4631-8D8F-2994853819F8', 0x20690000010405000000400AD7A3A335144186EB51B86FA15141703D0AD7CE361441D8A370ED24A1514130333333F33B1441D8A370CD51A15141C0F5285CE9381441E27A141E95A15141400AD7A3A335144186EB51B86FA1514101000000020000000001000000FFFFFFFF0000000003, 'SSMS')
```
2.)  I then opened ArcPro, selected the aforementioned newly created Facility, moved it from Box Elder county to Tooele county, and saved my edits:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/3b33030b-9714-493f-b2a4-b89578a24ab9)
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/34e98b18-0341-447b-a294-432d042d23b1)

3.) I then checked the FacilityID field and it had changed from `test` to `UTU03F170C8B36`
Here is a side by side comparison of facility polygon metadata before and after moving:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/73fd34b7-ee67-460e-ae34-cf3ec9803571)

**Comment by steveoh** [comment]
That seems like a problem.

**Comment by rsparker-utah** [comment]
Agreed. I did two more tests to add some additional context:

**Test 1**
1.) Within ArcPro, I selected a polygon that was created natively within ArcPro, moved it from Box Elder county to Tooele county, and saved my edits:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/801e6fd7-1120-4fc9-bce1-244739dfcac7)
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/86952b37-3605-4263-a778-590290d09676)

2.)  I then checked the FacilityID field and it had stayed the same.
Here is the side by side comparison of facility polygon metadata before and after moving:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/2bdb151e-c47e-4878-a2f1-59faddbae533)

**Test 2**
1.) Within ArcPro, I selected the polygon that was previously created in SSMS (this was the polygon that was originally created with a facilityID of "test", but ended up changing after being moved from Box Elder county to Tooele county within ArcPro). 

2.) I then moved this polygon back to Box Elder county from Tooele county, and saved my edits:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/c814749d-d675-4d43-a1c1-3ba02f6e6339)
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/746eab31-b08d-4623-9f01-094780d200c2)

3.) I then checked the FacilityID field and it had stayed the same.
Here is the side by side comparison of facility polygon metadata before and after moving:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/6ece4c3f-6974-427d-894c-4c82f0bdc583)

It appears there is some requirement that is not met/recognized by ArcPro when a facility polygon record is first created in SSMS, which then causes ArcPro to initiate the FacilityID creation attribute rule the first time the polygon is moved.

**Comment by nathankota** [comment]
Perhaps a simplistic view here, but if we're wanting to rely on the app to generate FacilityID from now on, can the attribute rule for it be retired?

**Comment by rsparker-utah** [comment]
This sounds reasonable to me. Assuming we went with this approach, would we have the ability to manually enter a FacilityID in ArcPro as a backup in case the FacilityID generation with the web app was down?

**Comment by nathankota** [comment]
> ...would we have the ability to manually enter a FacilityID in ArcPro as a backup in case the FacilityID generation with the web app was down?

My assumption is yes you would be able, and in fact required, to manually edit the FacilityID in the case where it wasn't populated by the app for some reason; or if you needed to manually enter an inventory like you do now; or if you needed to change a FacilityID for some reason. @steveoh do you agree with this?

My follow-up question is: What are the other ramifications of possibly removing this attribute rule, and are you willing to accept them?

**Comment by rsparker-utah** [comment]
Ok, that's good to hear that we would be able to enter the FacilityID manually as well if needed. Assuming removing the attribute rule will make the the FacilityID field editable across the board, the only other potential issue/ramification I can think of would be that now there is the potential that an established FacilityID could accidentally be edited/changed (currently the FacilityID field becomes a non-editable field after creation). The FacilityID has been the unique identifier provided to applicants (via the ABR letter) for their submission, so changes to existing records could potentially be an issue. Do we have any options around this? For example, is there a way to have the FacilityID be editable during the facility polygon creation, but then make it a non-editable field once the facility polygon has been established? Or something to that effect?

**Comment by steveoh** [comment]
Offhand I don't know the answer. 

Research, testing, and esri support might be necessary to figure that out definitively. You have much more experience than I do with SDE. I don't understand the processes and data entry methods well enough to know how easily the id could change and go unnoticed. Would your versioning model catch that type of error?

When the only tool you have is a hammer my first hunch is to use some sort of attribute rule.

**Comment by rsparker-utah** [comment]
My understanding of the versioning process is that if the FacilityID was accidentally changed within one version, then it would be updated in all of the versions after the reconcile and post process. 

That being said, after speaking to my colleagues and further consideration, we are fine with retiring the attribute rule for the FacilityID in the geodatabase/ArcPro.

**Comment by nathankota** [comment]
@rsparker-utah What do you think about changing the F to S in the ID since we mostly refer to the parent location as Site rather than Facility in the app? It will also provide a quick visual identification of records in the database that came from the app versus those that were pre-app.

**Comment by rsparker-utah** [comment]
@nathankota I ran this by our UIC group and we all thought this makes sense. We are for this idea.

**Comment by steveoh** [comment]
@nathankota this is 7 requests bundled up in one. I'll take the liberty to add the field to the report and review page where it seems to fit. Is the ARB letter still relevant? The last one is probably handled by another issue. Do you agree?

**Comment by nathankota** [comment]
> @nathankota this is 7 requests bundled up in one. 
I struggled with lumping vs splitting, and went with the former to keep it all together and tight. If it doesn't work as well for you, I will use it as a lesson for the future.   

> I'll take the liberty to add the field to the review page where it seems to fit. 
👍 

> The report page is the same page printed right? 
Yes, with layout updates eventually via #273 

> Is the ARB letter still relevant? 
Yes ref: #282 

> The last one is probably handled by another issue. Do you agree?
I'll put it in #284 so it doesn't get lost, and remove it from here.

**Comment by nathankota** [comment]
@rsparker-utah, I tested this a bit and it looks good to me, but I'd appreciate your further validation that it is ready to close. In particular, I kept an eye on the ID format, particularly FIPS.

**Comment by nathankota** [comment]
@rsparker-utah please confirm that this works as expected

**Comment by rsparker-utah** [comment]
Tested and validated on the following sites/inventories: 
[Site 109, inventory 201](https://uic-inventory.dev.utah.gov/review/site/109/inventory/201)
[Site 110, inventory 202](https://uic-inventory.dev.utah.gov/review/site/110/inventory/202)
[Site 111, inventory 203](https://uic-inventory.dev.utah.gov/review/site/111/inventory/203)
